refactor(core): make the error messages tree shakable

ramthir opened the pull request:
Long error messages can be tree-shaken in the production build and replaced with error codes.

See: https://github.com/angular/angular/pull/44219#issuecomment-983216374

## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->

- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [x] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:


## What is the current behavior?
Long error messages are not tree-shakable in the production build.

Issue Number: N/A


## What is the new behavior?
Long error messages are tree-shaken in the production build and replaced with error codes.

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No


<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. -->


## Other information

ramthir committed:
refactor(core): make the error messages tree shakable
Long error messages can be tree-shaken in the production build and replaced with error codes.

See: https://github.com/angular/angular/pull/44219#issuecomment-983216374

ramthir commented on packages/core/src/render3/error_code.ts:
```diff
@@ -42,6 +47,16 @@ export const enum RuntimeErrorCode {
   // Compilation Errors
+
+  // Other
+  INVALID_ARGUMENTS = '900',
```
@AndrewKushnir I need some help with the error codes, not sure about the sections of these and if we can group some of them.

AndrewKushnir replied:
@ramthir thanks for the comment, I can provide some feedback during the review process.

AndrewKushnir replied:
I've left some comments and proposals on the error codes. Could you please apply the changes and I'll take a look at the final list to figure out if we can re-group the new ones? Thank you.

AndrewKushnir commented on packages/core/src/change_detection/differs/default_iterable_differ.ts:
```diff
@@ -152,8 +153,10 @@ export class DefaultIterableDiffer<V> implements IterableDiffer<V>, IterableChan
+      const errorMessage = (typeof ngDevMode === 'undefined' || ngDevMode) ?
+          `Error trying to diff '${stringify(collection)}'. Only arrays and iterables are allowed` :
+          '';
+      throw new RuntimeError(RuntimeErrorCode.INVALID_ARGUMENTS, errorMessage);
```
```suggestion
      throw new RuntimeError(RuntimeErrorCode.INVALID_DIFFER_INPUT, errorMessage);
```

AndrewKushnir commented on packages/core/src/change_detection/differs/default_keyvalue_differ.ts:
```diff
@@ -79,8 +81,10 @@ export class DefaultKeyValueDiffer<K, V> implements KeyValueDiffer<K, V>, KeyVal
+      const errorMessage = (typeof ngDevMode === 'undefined' || ngDevMode) ?
+          `Error trying to diff '${stringify(map)}'. Only maps and objects are allowed` :
+          '';
+      throw new RuntimeError(RuntimeErrorCode.INVALID_ARGUMENTS, errorMessage);
```
```suggestion
      throw new RuntimeError(RuntimeErrorCode.INVALID_DIFFER_INPUT, errorMessage);
```

AndrewKushnir commented on packages/core/src/change_detection/differs/iterable_differs.ts:
```diff
@@ -250,8 +251,11 @@ export class IterableDiffers {
+          `Cannot find a differ supporting object '${iterable}' of type '${
+              getTypeNameForDebugging(iterable)}'` :
+          '';
+      throw new RuntimeError(RuntimeErrorCode.NO_SUPPORTING_FACTORY, errorMessage);
```
```suggestion
      throw new RuntimeError(RuntimeErrorCode.NO_SUPPORTING_DIFFER_FACTORY, errorMessage);
```

AndrewKushnir commented on packages/core/src/change_detection/differs/keyvalue_differs.ts:
```diff
@@ -181,6 +183,9 @@ export class KeyValueDiffers {
+    const errorMessage = (typeof ngDevMode === 'undefined' || ngDevMode) ?
+        `Cannot find a differ supporting object '${kv}'` :
+        '';
+    throw new RuntimeError(RuntimeErrorCode.NO_SUPPORTING_FACTORY, errorMessage);
```
```suggestion
    throw new RuntimeError(RuntimeErrorCode.NO_SUPPORTING_DIFFER_FACTORY, errorMessage);
```

AndrewKushnir commented on packages/core/src/di/injector_compatibility.ts:
```diff
@@ -98,15 +102,15 @@ export function ɵɵinject<T>(token: ProviderToken<T>, flags = InjectFlags.Defau
  */
 export function ɵɵinvalidFactoryDep(index: number): never {
-  const msg = ngDevMode ?
+  const errorMessage = (typeof ngDevMode === 'undefined' || ngDevMode) ?
```
Since this is a part of the generated code, I think we should avoid using extra classes here. Let's revert to the previous state.

AndrewKushnir commented on packages/core/src/di/injector_compatibility.ts:
```diff
@@ -58,7 +59,10 @@ export function injectInjectorOnly<T>(token: ProviderToken<T>, flags?: InjectFla
+    const errorMessage = typeof ngDevMode === 'undefined' || ngDevMode ?
+        `inject() must be called from an injection context` :
+        '';
+    throw new RuntimeError(RuntimeErrorCode.OUTSIDE_INJECTION_CONTEXT, errorMessage);
```
```suggestion
    throw new RuntimeError(RuntimeErrorCode.MISSING_INJECTION_CONTEXT, errorMessage);
```

AndrewKushnir commented on packages/core/src/di/injector_compatibility.ts:
```diff
@@ -98,15 +102,15 @@ export function ɵɵinject<T>(token: ProviderToken<T>, flags = InjectFlags.Defau
-      'invalid';
-  throw new Error(msg);
+      '';
+  throw new RuntimeError(RuntimeErrorCode.INVALID_FACTORY_DEPENDENCY, errorMessage);
```
Same as above, let's revert this change (keep the code as is).

AndrewKushnir commented on packages/core/src/di/injector_compatibility.ts:
```diff
@@ -141,7 +145,10 @@ export function injectArgs(types: (ProviderToken<any>|any[])[]): any[] {
+        const errorMessage = (typeof ngDevMode === 'undefined' || ngDevMode) ?
+            'Arguments array must have arguments.' :
+            '';
+        throw new RuntimeError(RuntimeErrorCode.INVALID_ARGUMENTS, errorMessage);
```
```suggestion
        throw new RuntimeError(RuntimeErrorCode.INVALID_INJECTION_ARGUMENTS, errorMessage);
```

AndrewKushnir commented on packages/core/src/render3/error_code.ts:
```diff
@@ -72,9 +87,9 @@ export function formatRuntimeError(code: RuntimeErrorCode, message: string): str
-  // there instead to tree-shake more devmode-only code (and eventually remove `ngDevMode` check
-  // from this code).
+  // TODO: revisit all instances where `RuntimeError` is thrown and see if `ngDevMode` can be
+  // added there instead to tree-shake more devmode-only code (and eventually remove `ngDevMode`
+  // check from this code).
```
This change looks unnecessary, could you please revert?

AndrewKushnir commented on packages/core/src/render3/features/inherit_definition_feature.ts:
```diff
@@ -39,7 +40,10 @@ export function ɵɵInheritDefinitionFeature(definition: DirectiveDef<any>|Compo
+        const errorMessage = typeof ngDevMode === 'undefined' || ngDevMode ?
+            'Directives cannot inherit Components' :
+            '';
+        throw new RuntimeError(RuntimeErrorCode.DIRECTIVES_CANNOT_INHERIT_COMPONENTS, errorMessage);
```
```suggestion
        throw new RuntimeError(RuntimeErrorCode.INVALID_INHERITANCE, errorMessage);
```

AndrewKushnir commented on packages/core/src/render3/i18n/i18n_apply.ts:
```diff
@@ -198,7 +200,10 @@ export function applyMutableOpCodes(
           break;
         default:
-          throw new Error(`Unable to determine the type of mutate operation for "${opCode}"`);
+          const errorMessage = typeof ngDevMode === 'undefined' || ngDevMode ?
+              `Unable to determine the type of mutate operation for "${opCode}"` :
+              '';
+          throw new RuntimeError(RuntimeErrorCode.MUTABLE_OPCODES, errorMessage);
```
In this code we can call `ngDevMode` without extra checks (see some usages above). We could also wrap the entire throw into `ngDevMode` check (it's ok to tree-shake away this in prod bundles):
```suggestion
          if (ngDevMode) {
            throw new RuntimeError(
              RuntimeErrorCode.INVALID_I18N_STRUCTURE,
              `Unable to determine the type of mutate operation for "${opCode}"`);
          }
```

AndrewKushnir commented on packages/core/src/render3/view_ref.ts:
```diff
@@ -284,7 +286,10 @@ export class ViewRef<T> implements viewEngine_EmbeddedViewRef<T>, viewEngine_Int
   attachToViewContainerRef() {
     if (this._appRef) {
-      throw new Error('This view is already attached directly to the ApplicationRef!');
+      const errorMessage = typeof ngDevMode === 'undefined' || ngDevMode ?
+          'This view is already attached directly to the ApplicationRef!' :
+          '';
+      throw new RuntimeError(RuntimeErrorCode.VIEW_ALREADY_ATTACHED, errorMessage);
```
It should be safe to call `ngDevMode` without extra checks here as well:
```suggestion
      const errorMessage = ngDevMode ?
          'This view is already attached directly to the ApplicationRef!' :
          '';
      throw new RuntimeError(RuntimeErrorCode.VIEW_ALREADY_ATTACHED, errorMessage);
```

AndrewKushnir commented on packages/core/src/render3/view_ref.ts:
```diff
@@ -296,7 +301,10 @@ export class ViewRef<T> implements viewEngine_EmbeddedViewRef<T>, viewEngine_Int
   attachToAppRef(appRef: ViewRefTracker) {
     if (this._attachedToViewContainer) {
-      throw new Error('This view is already attached to a ViewContainer!');
+      const errorMessage = typeof ngDevMode === 'undefined' || ngDevMode ?
+          'This view is already attached to a ViewContainer!' :
+          '';
+      throw new RuntimeError(RuntimeErrorCode.VIEW_ALREADY_ATTACHED, errorMessage);
```
```suggestion
      const errorMessage = ngDevMode ?
          'This view is already attached to a ViewContainer!' :
          '';
      throw new RuntimeError(RuntimeErrorCode.VIEW_ALREADY_ATTACHED, errorMessage);
```

AndrewKushnir commented on packages/core/src/sanitization/sanitization.ts:
```diff
@@ -117,7 +118,10 @@ export function ɵɵsanitizeResourceUrl(unsafeResourceUrl: any): TrustedScriptUR
     return trustedScriptURLFromStringBypass(unwrapSafeValue(unsafeResourceUrl));
   }
-  throw new Error('unsafe value used in a resource URL context (see https://g.co/ng/security#xss)');
+  const errorMessage = typeof ngDevMode === 'undefined' || ngDevMode ?
```
I'd propose keeping parenthesis around the condition here (and in other cases in this PR where parenthesis are missing) for better readability:
```suggestion
  const errorMessage = (typeof ngDevMode === 'undefined' || ngDevMode) ?
```

AndrewKushnir commented on packages/core/src/sanitization/sanitization.ts:
```diff
@@ -165,7 +171,10 @@ export function ɵɵtrustConstantHtml(html: TemplateStringsArray): TrustedHTML|s
   // TemplateStringsArray is 1.
   if (ngDevMode && (!Array.isArray(html) || !Array.isArray(html.raw) || html.length !== 1)) {
-    throw new Error(`Unexpected interpolation in trusted HTML constant: ${html.join('?')}`);
+    const errorMessage = typeof ngDevMode === 'undefined' || ngDevMode ?
+        `Unexpected interpolation in trusted HTML constant: ${html.join('?')}` :
+        '';
+    throw new RuntimeError(RuntimeErrorCode.UNEXPECTED_INTERPOLATION, errorMessage);
```
This code is already in the `ngDevMode` check, so you can drop the check and just use error message as is.

AndrewKushnir commented on packages/core/src/sanitization/sanitization.ts:
```diff
@@ -191,7 +200,10 @@ export function ɵɵtrustConstantResourceUrl(url: TemplateStringsArray): Trusted
   // the length of the TemplateStringsArray is 1.
   if (ngDevMode && (!Array.isArray(url) || !Array.isArray(url.raw) || url.length !== 1)) {
-    throw new Error(`Unexpected interpolation in trusted URL constant: ${url.join('?')}`);
+    const errorMessage = typeof ngDevMode === 'undefined' || ngDevMode ?
+        `Unexpected interpolation in trusted URL constant: ${url.join('?')}` :
+        '';
+    throw new RuntimeError(RuntimeErrorCode.UNEXPECTED_INTERPOLATION, errorMessage);
```
Same here, we are in the `ngDevMode` block, so we can use the error message as is.

AndrewKushnir commented on packages/core/src/sanitization/sanitization.ts:
```diff
@@ -234,19 +246,23 @@ export function ɵɵsanitizeUrlOrResourceUrl(unsafeUrl: any, tag: string, prop:
-        `\nIf '${name}' is a directive input, make sure the directive is imported by the` +
-        ` current module.`;
-    throw new Error(msg);
+    const errorMessage = typeof ngDevMode === 'undefined' || ngDevMode ?
```
Both functions below are called from within the blocks that are guarded by `ngDevMode`, so these functions are tree-shaken away already, so we should only add the RuntimeError class here.

ramthir replied:
Yeah, I'll do that. Also will keep a look on this in future changes. 👍

AndrewKushnir commented on packages/core/src/render3/error_code.ts:
```diff
@@ -42,6 +47,16 @@ export const enum RuntimeErrorCode {
+  NO_SUPPORTING_FACTORY = '901',
+  VIEW_ALREADY_ATTACHED = '902',
+  DIRECTIVES_CANNOT_INHERIT_COMPONENTS = '903',
+  MUTABLE_OPCODES = '904',
```
I think this one should go under the `i18n Errors` section.

AndrewKushnir reviewed (commented):
@ramthir thanks for creating this PR 👍 I've left some comments, could you please take a look when you get a chance?

Also, according to the CI, this change brings ~600 bytes improvement 🎉

AndrewKushnir commented on packages/core/src/render3/error_code.ts:
```diff
@@ -40,8 +45,18 @@ export const enum RuntimeErrorCode {
+  INVALID_INHERITANCE = '903',
+  UNSAFE_VALUE_IN_RESOURCE_URL = '904',
+  UNSAFE_VALUE_IN_SCRIPT = '905',
+  UNEXPECTED_INTERPOLATION = '906'
```
It looks like this code is unused.
```suggestion
```

AndrewKushnir reviewed (commented):
@ramthir thanks for addressing the feedback, the grouping for the error codes look good (there is just 1 comment, see below). Could you please change the status of the PR from Draft -> Ready for Review, so that we complete the review? Thank you.

ramthir commented:
@AndrewKushnir, I've made the requested changes. Would you please let me know if I need to make any other changes to this PR?

AndrewKushnir commented:
@ramthir thanks for the updates. The changes look good and I'll have another look once PR #44348 lands (this PR would require some minor changes after that). I will keep you updated. Thank you.

AndrewKushnir commented:
@ramthir just a quick update: we are in the process of discussing an update to the error code format to allow 5 digit codes (currently only 4 digit codes are supported) in a context of https://github.com/angular/angular/pull/44398. It would probably take some time, but we'd need to resolve that before we can proceed with this PR. There are no action items required at this moment, I just wanted to let you know that we'll get back to this PR once https://github.com/angular/angular/pull/44398 is resolved. Thank you.

ramthir commented:
Thanks for taking the time to update @AndrewKushnir. Looking forward to this change, we can proceed in other areas. 

In the meantime, please let me know if there are any other areas we can improve on while waiting on this.

josephperrott reviewed (approved):
LGTM

AndrewKushnir commented:
@ramthir just wanted to let you know that PR https://github.com/angular/angular/pull/44398 is now merged into the main branch and this PR is no longer blocked. The list of error code has moved to `packages/core/src/errors.ts` and error codes are now integers (vs strings). Could you please rebase this PR and update the code according to the current state? Please let us know if you have any questions. Thank you.

ramthir commented:
@AndrewKushnir I have a test failure due to golden with `stringify6`; how do we proceed?

AndrewKushnir commented:
> I have a test failure due to golden with stringify6; how do we proceed?

@ramthir please run the following command locally - it'd update a golden file as needed:

```
yarn bazel run //packages/core/test/bundling/router:symbol_test.accept
```

AndrewKushnir reviewed (approved):
@ramthir thanks for additional updates! The changes look good 👍

AndrewKushnir reviewed (approved):
Reviewed-for: public-api, size-tracking

AndrewKushnir commented:
[Presubmit](http://test/OCL:421917265:BASE:421917303:1642200444894:e07643ad).

atscott reviewed (approved):
reviewed-for: public-api
reviewed-for: size-tracking

dylhunn reviewed (approved):
reviewed-for: public-api

alxhub reviewed (approved):
Reviewed-for: global-approvers

Two questions:

1) is it intentional that some error messages are guarded with a plain read of `ngDevMode` and others use a `typeof` guard before reading `ngDevMode`?
2) what do you think about altering the constructor of `RuntimeError` to allow for the `false` from `ngDevMode && 'error message'` to be passed through directly? That would eliminate the need for a temporary variable for the error string, as well as the need to use `''` explicitly in those cases.

AndrewKushnir commented:
@alxhub thanks for the review 👍 

> 1. is it intentional that some error messages are guarded with a plain read of ngDevMode and others use a typeof guard before reading ngDevMode?

This is a ViewEngine artifact, we should be able to get rid of it, but it'd need some investigation (to see if this would be a breaking change for people that build their apps by custom tooling (outside of our CLI)).

> 2. what do you think about altering the constructor of RuntimeError to allow for the false from ngDevMode && 'error message' to be passed through directly? That would eliminate the need for a temporary variable for the error string, as well as the need to use '' explicitly in those cases.

That's an awesome idea 👍 I will update the `RuntimeError` class to support that and we can do more refactoring once we confirm that it's safe to use the `ngDevMode` flag without extra `typeof` checks.

AndrewKushnir commented:
@ramthir I think we can proceed with the PR in its current state and I'll let you know once there are remaining pieces in place to do further refactoring as Alex proposed.

AndrewKushnir committed:
fixup! refactor(core): make the error messages tree shakable

Long error messages can be tree-shaken in the production build and replaced with error codes.

AndrewKushnir commented:
@ramthir I've rebased this PR (there was 1 minor merge conflict) and this PR is now in the merge queue!

I wanted to say thanks for your help to standardize the error messages as well as making them more tree-shakable 👍

dylhunn commented:
This PR was merged into the repository by commit bc03a2e3e3add26a0c52e1ae6eba8ca5f5da5786.
